cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java

```diff
@@ -670,6 +738,221 @@ public void testSearchMetrics() throws Exception {
+                          ImmutableMap<String, ImmutableList<ImmutableList<QueryResult>>> expected) throws Exception {
+    String url = "/v3/metrics/query";
+    HttpURLConnection urlConn = openURL(MetricsSuiteTestBase.getEndPoint(url).toURL(),
+                                        HttpMethod.POST);
+    urlConn.setDoOutput(true);
+    String jsonBatch = new Gson().toJson(batchRequests);
+    urlConn.getOutputStream().write(jsonBatch.getBytes());
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(), urlConn.getResponseCode());
+
+    String result = IOUtils.toString(urlConn.getInputStream(), Charsets.UTF_8);
+    Map<String, List<MetricQueryResult>> results =
+      new Gson().fromJson(result, new TypeToken<Map<String, List<MetricQueryResult>>>() {
+      }.getType());
+    urlConn.disconnect();
+    // check we have all the keys
+    Assert.assertEquals(expected.keySet(), results.keySet());
+    for (Map.Entry<String, List<MetricQueryResult>> entry : results.entrySet()) {
+      ImmutableList<ImmutableList<QueryResult>> expectedQueryResult = expected.get(entry.getKey());
+      List<MetricQueryResult> actualQueryResult = entry.getValue();
+      // check list length size
+      Assert.assertEquals(expectedQueryResult.size(), actualQueryResult.size());
+      for (int i = 0; i < expectedQueryResult.size(); i++) {
+        compareQueryResults(expectedQueryResult.get(i), actualQueryResult.get(i));
+      }
+    }
+  }
+
+  private void compareQueryResults(ImmutableList<QueryResult> expected, MetricQueryResult actual) {
+    MetricQueryResult.TimeSeries[] actualTimeSeries = actual.getSeries();
+    sortTimeSeriesBasedOnGrouping(actualTimeSeries);
+    // check length
+    Assert.assertEquals(expected.size(), actualTimeSeries.length);
+    for (int i = 0; i < expected.size(); i++) {
+
+      Assert.assertEquals(expected.get(i).getMetricName(), actualTimeSeries[i].getMetricName());
+      Assert.assertEquals(expected.get(i).getGrouping(), actualTimeSeries[i].getGrouping());
+      MetricQueryResult.TimeValue[] values = actualTimeSeries[i].getData();
+      long numPoints = 0;
+      long totalSum = 0;
+      for (MetricQueryResult.TimeValue tv : values) {
+        numPoints++;
+        totalSum += tv.getValue();
+      }
+      Assert.assertEquals(expected.get(i).getNumPoints(), numPoints);
+      Assert.assertEquals(expected.get(i).getTotalSum(), totalSum);
+    }
+  }
+
+  private void sortTimeSeriesBasedOnGrouping(MetricQueryResult.TimeSeries[] actualTimeSeries) {
+     Arrays.sort(actualTimeSeries, new Comparator<MetricQueryResult.TimeSeries>() {
+      @Override
+      public int compare(MetricQueryResult.TimeSeries t1, MetricQueryResult.TimeSeries t2) {
+        if (t1.getGrouping().size() == 0 && t2.getGrouping().size() == 0) {
+          return 0;
+        }
+        Map<String, String> t1Map = t1.getGrouping();
+        Map<String, String> t2Map = t2.getGrouping();
+        for (Map.Entry<String, String> entry1 : t1Map.entrySet()) {
```

Comment:
this is wrong, no? loop in a loop?

is there any other way without this sorting?